physics/discrete_trajectory_body.hpp

```diff
@@ -207,26 +204,26 @@ void DiscreteTrajectory<Frame>::ForgetAfter(Instant const& time) {
   // Get an iterator denoting the first entry with time > |time|.  Remove that
   // entry and all the entries that follow it.  This preserves any entry with
   // time == |time|.
-  auto const it = timeline_.upper_bound(time);
+  auto const first_removed_in_timeline = timeline_.upper_bound(time);
+  Instant const& first_removed_time = first_removed_in_timeline->first;
   if (downsampling_.has_value()) {
-    if (it != timeline_.end() &&
-        downsampling_->start_of_dense_timeline->first >= it->first) {
+    if (first_removed_in_timeline != timeline_.end() &&
+        downsampling_->first_dense_time() >= first_removed_time) {
```

Comment:
I still don't understand this code.  Let's talk.

Reply:
Done.

Reply:
Also added a test.